src/dune_engine/cached_digest.ml

```diff
@@ -148,43 +173,48 @@ let peek_file fn =
     Some
       (if x.stats_checked = cache.checked_key then
         x.digest
-      else (
-        needs_dumping := true;
-        let stat = Path.stat fn in
-        let dirty = ref false in
-        set_max_timestamp cache stat;
-        if stat.st_mtime <> x.timestamp then (
-          dirty := true;
-          x.timestamp <- stat.st_mtime
-        );
-        if stat.st_perm <> x.permissions then (
-          dirty := true;
-          x.permissions <- stat.st_perm
-        );
-        if stat.st_size <> x.size then (
-          dirty := true;
-          x.size <- stat.st_size
-        );
-        if !dirty then x.digest <- Digest.file_with_stats fn stat;
-        x.stats_checked <- cache.checked_key;
-        x.digest
-      ))
+      else
+        let stats = Path.stat fn in
+        let reduced_stats = Reduced_stats.of_unix_stats stats in
+        match Reduced_stats.compare x.stats reduced_stats with
+        | Eq ->
+          x.stats_checked <- cache.checked_key;
```

Comment:
It seems worth clarifying here why it's safe to skip `needs_dumping := true` even though you're modifying the data structure.
Maybe like this:

(* Even though we're modifying the `stats_checked` field, we don't need to set `needs_dumping := true` here. This is because `checked_key` is incremented every time we load from disk, which makes it so that all `stats_checked < checked_key` for all entries after loading, regardless of if we save the new value here or not. *)

Reply:
Good point. Thanks for pointing that out, I added your comment.